### ISSUE TITLE
Bootstrap Write error conditions

### ISSUE THREAD

**Issue by kFYatek** [OpenMobileAlliance/OMA_LwM2M_for_Developers#184]
The description below is based on the current draft 1.0 version of the LwM2M specification (OMA-TS-LightweightM2M-V1_0-20170111-D).

Section 5.2.7.1 BOOSTRAP WRITE (side note: there's a typo in the section header, it's missing the "T" from "BOO-T-STRAP") states that:
> When the “Write” operation targets an Object or an Object Instance, the LwM2M Client MUST ignore optional resources it does not support in the payload.

It only specifies to ignore "optional resources" that are not supported by the Client, but does not address other cases of dubious writes. For example, what are the expected outcomes of the following operations:
- BOOTSTRAP WRITE /3/0 (Device Object) with content including resource 2 (Serial Number)? - as an example of a mandatory strictly read-only resource
- BOOTSTRAP WRITE /4/0 (Connectivity Monitoring Object) with content including resource 3 (Link Quality)? - as above, but the resource is defined as optional
- BOOTSTRAP WRITE /3/0 (Device Object) with content including resource 4 (Reboot)? - as an example of an executable resource
- BOOTSTRAP WRITE /3/0 (Device Object) with content including resource 42? - as an example of an invalid resource number, never specified as neither mandatory nor optional
- BOOTSTRAP WRITE /1/1 (Server Object) with content including resource 7 (Binding) with value `invalid`? - as an example of trying to write a value outside of the domain of valid values for a given (itself valid) resource
- BOOTSTRAP WRITE /1/1 (Server Object) with content including resource 5 (Disable Timeout) with value `-69`? - similar to above, but the resource is defined as optional

Moreover, I used the term "strictly read-only resource", even though it does not exist in the specification. The ability to write to a resource is never strictly defined. For example:
- Resource /1/*/0 (LwM2M Server / Short Server ID) has "R" specified in the Operations field, making it read-only for regular servers, even though it obviously shall be actually writeable by the Bootstrap Server
- Resource /4/0/3 (LwM2M Connectivity Monitoring / Link Quality) also has "R" specified in the Operations field, making it read-only for regular servers, and it obviously is impossible to write it by the Bootstrap Server, either

This distinction can only be made through intuition, while I believe it should be explicitly specified.

To summarize: while performing a Bootstrap Write operation targeting an Object or an Object Instance, when exactly to silently ignore a resource entry? When to return an error code? And how to determine which nominally read-only resources may in fact be written to by the Bootstrap Server?

**Comment by dnav** [comment]
Hi,

A point that may not be clear enough in the TS: the operation restrictions are for LWM2M Servers only. They must be ignored for the Bootstrap Interface. See appendix D.1:

> **Operations**: indicates which operations the Resource supports in the “Device Management & Service Enablement” Interface. This field can be set to a combination of R (Read, Observe, Discover, Write-Attributes), and W (Write), or can be set to E (Execute); Executable Operation is exclusive regarding the two others (R,W). This field may also have an empty value, which means that this field is not allowed to be accessed via “Device Management & Service Enablement” Interface but allowed to be accessed via “Bootstrap” Interface.

Obviously, the executable resources does not need to be present in the Bootstrap payload.

Regards,

**Comment by kFYatek** [comment]
Yes. This is all fairly obvious. But it doesn't specify what the Client is supposed to do upon receiving such Bootstrap Write requests. It also doesn't specify which resources are writeable by the Bootstrap Server and which are not.

**Comment by dnav** [comment]
All resources are writable by the Bootstrap-Server. All of them. However the Write may fail because of the Client internal limitations.

If the Bootstrap Write targets a resource which is hard-coded in the device like the Serial Number. The operation is a valid one protocol-wise but as the Client can not perform the operation, it should return an error.
If the Bootstrap Write contains invalid values, the Client should return an error.
If it contains unknown IDs, it is the same case as resources not supported by the Client. You can not expect the Client to be aware of all optional resources IDs.

Regards,

**Comment by kFYatek** [comment]
So, to sum up:
* If the Client supports a given resource, but is unable to perform the write operation (because it is e.g. hard-coded like Serial Number or the value is invalid), then it shall return an error, regardless of whether the operation targets the Resource itself or enclosing Instance or Object.
  * Is it the same when the given resource is Optional, such as the afore-mentioned Link Quality (likely a transient value read from the driver)?
* If a resource is not supported by the Client, it shall be ignored when it is written as part of an operation targeting the Instance or Object.

Is this right?

**Comment by dnav** [comment]
Yes.

Regards,

**Comment by jpradocueva** [comment]
Agreed with DM Working Group to close.
